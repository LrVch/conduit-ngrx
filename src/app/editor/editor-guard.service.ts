import { Injectable } from '@angular/core';
import {
    CanActivate,
    ActivatedRouteSnapshot
} from '@angular/router';
import { Observable, of, throwError } from 'rxjs';
import { Store, select } from '@ngrx/store';
import { AppState } from '@app/reducers';
import { Router } from '@angular/router';
import { tap, take, map, switchMap, catchError, withLatestFrom, filter, mergeMap } from 'rxjs/operators';
import { selectArticle } from './editor.selectors';
import { ArticlesService } from '@app/core';
import { EditorArticleLoadSuccess } from './editor.actions';
import { selectUser } from '@app/auth/auth.selectors';


@Injectable()
export class EditorGuard implements CanActivate {
    constructor(
        private store: Store<AppState>,
        private router: Router,
        private articlesService: ArticlesService
    ) {

    }
    canActivate(
        route: ActivatedRouteSnapshot
    ): Observable<boolean> {

        const slug = route.params['slug'];

        return this.store.pipe(
            take(1),
            switchMap(() => this.articlesService.get(slug)),
            withLatestFrom(this.store.select(selectUser)),
            mergeMap(([article, user]) => article.author.username === user.username
            ? of(new EditorArticleLoadSuccess({ article })) : throwError('wrong user')),
            tap((action: EditorArticleLoadSuccess) => this.store.dispatch(action)),
            map(article => !!article),
            catchError((error) => {
                this.router.navigate(['/']);
                console.error(error);
                return of(false);
            })
        );
    }
}
