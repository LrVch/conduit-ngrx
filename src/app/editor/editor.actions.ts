import { Action } from '@ngrx/store';
import { Article, Errors } from '@app/core';

export enum EditorActionTypes {
  EditorArticleLoadSuccess = '[Editor] Editor Article Load Success',
  EditorArticleLoadFail = '[Editor] Editor Article Load Fail',
  EditorArticleClear = '[Editor] Editor Article Clear',

  EditorArticleSaveRequest = '[Editor] Editor Article Save Request',
  EditorArticleSaveSuccess = '[Editor] Editor Article Save Success',
  EditorArticleSaveFail = '[Editor] Editor Article Save Fail',

  ClearEditorErrors = '[Editor] Clear Editor Errors',
  SetEditorErrors = '[Editor] Set Editor Errors'
}

export class EditorArticleLoadSuccess implements Action {
  readonly type = EditorActionTypes.EditorArticleLoadSuccess;

  constructor(public payload: { article: Article }) {}
}

export class EditorArticleClear implements Action {
  readonly type = EditorActionTypes.EditorArticleClear;
}

export class EditorArticleSaveRequest implements Action {
  readonly type = EditorActionTypes.EditorArticleSaveRequest;

  constructor(public payload: { article: Article }) {}
}

export class EditorArticleSaveSuccess implements Action {
  readonly type = EditorActionTypes.EditorArticleSaveSuccess;

  constructor(public payload: { article: Article }) {}
}

export class EditorArticleSaveFail implements Action {
  readonly type = EditorActionTypes.EditorArticleSaveFail;

  constructor(public payload: { errors: Errors }) {}
}

export class ClearEditorErrors implements Action {
  readonly type = EditorActionTypes.ClearEditorErrors;
}

export class SetEditorErrors implements Action {
  readonly type = EditorActionTypes.SetEditorErrors;

  constructor(public payload: { errors: Errors }) {}
}

export type EditorActions =
  | EditorArticleLoadSuccess
  | EditorArticleClear
  | EditorArticleSaveRequest
  | EditorArticleSaveSuccess
  | EditorArticleSaveFail
  | SetEditorErrors
  | ClearEditorErrors;
