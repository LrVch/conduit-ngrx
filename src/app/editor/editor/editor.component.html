<div class="editor-page page">
  <div class="container ">
    <div class="row">
      <div class="col-md-10 offset-md-1 col-xs-12">

        <app-list-error [errors]="errors$ | async"></app-list-error>

        <div class="mb-4">
          <app-editor-form [article]="article$ | async"
                           [isSubmitting]="isSubmitting$ | async"
                           (onsubmit)="onSubmit($event)"></app-editor-form>
        </div>

      </div>
    </div>
  </div>
</div>