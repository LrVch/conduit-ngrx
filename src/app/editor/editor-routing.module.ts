import { NgModule } from '@angular/core';
import { Routes, RouterModule } from '@angular/router';
import { EditorComponent } from './editor/editor.component';
import { AuthGuard } from '../core';
import { EditorGuard } from './editor-guard.service';
import { CanDeactivateGuard } from '../core/services/can-deactivate.guard';

const routes: Routes = [
  {
    path: '',
    component: EditorComponent,
    canActivate: [AuthGuard],
    canDeactivate: [CanDeactivateGuard]
  },
  {
    path: ':slug',
    component: EditorComponent,
    canActivate: [AuthGuard, EditorGuard],
    canDeactivate: [CanDeactivateGuard]
  }
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class EditorRoutingModule {}
