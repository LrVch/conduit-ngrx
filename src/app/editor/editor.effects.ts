import { Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { ArticlesService } from '../core';
import { EditorActionTypes, EditorArticleSaveRequest, EditorArticleSaveSuccess, EditorArticleSaveFail } from './editor.actions';
import { switchMap, map, catchError, tap } from 'rxjs/operators';
import { of } from 'rxjs';
import { Router } from '@angular/router';
import { AppState } from '../reducers';
import { Store } from '@ngrx/store';
import { ShowMainLoader, HideMainLoader } from '../layout/layout.actions';


@Injectable()
export class EditorEffects {

  constructor(
    private actions$: Actions,
    private articleService: ArticlesService,
    private router: Router,
    private sotre: Store<AppState>
  ) { }

  @Effect()
  saveArticle$ = this.actions$.pipe(
    ofType<EditorArticleSaveRequest>(EditorActionTypes.EditorArticleSaveRequest),
    switchMap(action => this.articleService.save(action.payload.article).pipe(
      map(article => new EditorArticleSaveSuccess({ article })),
      catchError(errors => {
        console.error(errors);
        return of(new EditorArticleSaveFail({ errors }));
      })
    )));

  @Effect({dispatch: false})
  editorArticleSaveSuccess$ = this.actions$.pipe(
    ofType<EditorArticleSaveSuccess>(EditorActionTypes.EditorArticleSaveSuccess),
    map(action => action.payload.article),
    tap(article =>  this.router.navigateByUrl('/article/' + article.slug))
  );

  @Effect({dispatch: false})
  editorArticleSaveRequest$ = this.actions$.pipe(
    ofType<EditorArticleSaveRequest>(EditorActionTypes.EditorArticleSaveRequest),
    tap(_ => this.sotre.dispatch(new ShowMainLoader())),
  );

  @Effect({dispatch: false})
  editorArticleSaveFail$ = this.actions$.pipe(
    ofType<EditorArticleSaveFail>(EditorActionTypes.EditorArticleSaveFail),
    tap(_ => this.sotre.dispatch(new HideMainLoader()))
  );
}
