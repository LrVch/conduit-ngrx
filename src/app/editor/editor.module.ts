import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { EditorComponent } from './editor/editor.component';
import { SharedModule } from '../shared';
import { EditorRoutingModule } from './editor-routing.module';
import { EditorFormComponent } from './editor-form/editor-form.component';
import { StoreModule } from '@ngrx/store';
import * as fromEditor from './editor.reducer';
import { EffectsModule } from '@ngrx/effects';
import { EditorEffects } from './editor.effects';
import { EditorGuard } from './editor-guard.service';
import { TranslateModule, TranslateLoader } from '@ngx-translate/core';
import { TranslateHttpLoader } from '@ngx-translate/http-loader';
import { HttpClient } from '@angular/common/http';
import { environment } from '@env/environment';

@NgModule({
  imports: [
    CommonModule,
    SharedModule,
    EditorRoutingModule,
    StoreModule.forFeature('editor', fromEditor.editorReducer),
    EffectsModule.forFeature([EditorEffects]),
    TranslateModule.forChild({
      loader: {
        provide: TranslateLoader,
        useFactory: HttpLoaderFactory,
        deps: [HttpClient]
      }
    }),
  ],
  declarations: [EditorComponent, EditorFormComponent],
  providers: [EditorGuard]
})
export class EditorModule { }

export function HttpLoaderFactory(http: HttpClient) {
  return new TranslateHttpLoader(
    http,
    `${environment.i18nPrefix}/assets/i18n/`,
    '.json'
  );
}
