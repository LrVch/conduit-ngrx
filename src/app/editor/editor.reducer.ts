import { Action } from '@ngrx/store';
import { Article, Errors } from '../core';
import { EditorActionTypes, EditorActions } from './editor.actions';


export interface EditorState {
  article: Article;
  savingArticle: boolean;
  errors: Errors;
}

export const initialState: EditorState = {
  article: null,
  savingArticle: false,
  errors: {
    errors: {}
  }
};

export function editorReducer(state = initialState, action: EditorActions): EditorState {
  switch (action.type) {

    case EditorActionTypes.EditorArticleLoadSuccess:
    case EditorActionTypes.EditorArticleSaveSuccess:
      return {
        ...state,
        article: action.payload.article,
        savingArticle: false,
        errors: {
          errors: {}
        }
      };

    case EditorActionTypes.EditorArticleClear:
      return {
        ...state,
        article: null
      };
    case EditorActionTypes.EditorArticleSaveRequest:
      return {
        ...state,
        savingArticle: true,
        errors: {
          errors: {}
        }
      };
    case EditorActionTypes.EditorArticleSaveFail:
      return {
        ...state,
        savingArticle: false,
        errors: { ...action.payload.errors }
      };
    case EditorActionTypes.SetEditorErrors:
      return {
        ...state,
        errors: { ...action.payload.errors }
      };
    case EditorActionTypes.ClearEditorErrors:
      return {
        ...state,
        errors: {
          errors: {}
        }
      };

    default:
      return state;
  }
}
