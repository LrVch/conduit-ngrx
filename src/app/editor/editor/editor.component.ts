import { Component, OnInit, OnDestroy } from '@angular/core';
import { Article, Errors } from '@app/core';
import { Observable } from 'rxjs';
import { Store, select } from '@ngrx/store';
import { AppState } from '@app/reducers';
import { selectArticle, selectArticleSaving, selectEditorErrors } from '@app/editor/editor.selectors';
import { EditorArticleSaveRequest, EditorArticleClear, ClearEditorErrors } from '@app/editor/editor.actions';
import { CanComponentDeactivate } from '@app/core/services/can-deactivate.guard';
import { DialogService } from '@app/core/services/dialog.service';
import { TranslateService } from '@ngx-translate/core';
import { first, switchMap, map, tap, mergeAll } from 'rxjs/operators';
import { MatDialogRef } from '@angular/material';


export interface Tag {
  name: string;
}

@Component({
  selector: 'app-editor',
  templateUrl: './editor.component.html'
})
export class EditorComponent implements OnInit, OnDestroy, CanComponentDeactivate {
  errors$: Observable<Errors>;
  isSubmitting$: Observable<boolean>;
  article$: Observable<Article>;

  wasChanged = false;

  constructor(
    private store: Store<AppState>,
    private dialog: DialogService,
    private translateService: TranslateService
  ) { }

  ngOnInit() {
    this.article$ = this.store.pipe(select(selectArticle));
    this.isSubmitting$ = this.store.pipe(select(selectArticleSaving));
    this.errors$ = this.store.pipe(select(selectEditorErrors));

    this.store.dispatch(new ClearEditorErrors());
  }

  ngOnDestroy() {
    this.store.dispatch(new EditorArticleClear());
  }

  onSubmit(article: Article) {
    this.store.dispatch(new EditorArticleSaveRequest({ article }));
  }

  onWasChanged(wasChanged: boolean) {
    this.wasChanged = wasChanged;
  }

  canDeactivate(): Observable<boolean> | boolean {
    if (!this.wasChanged) {
      return true;
    }

    return this.translateService.get('conduit.editor.goAwayWarning').pipe(
      map(question => this.dialog.confirmation({
        data: { question: question },
      })),
      switchMap(ref => ref.afterClosed()),
    );
  }
}
