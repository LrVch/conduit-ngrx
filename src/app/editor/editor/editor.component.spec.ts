import { async, ComponentFixture } from '@angular/core/testing';
import { ConfigureFn, configureTests, getArticle } from '../../lib/testing';
import { EditorComponent } from './editor.component';
import { DebugElement, Component, Output, EventEmitter, Input } from '@angular/core';
import { NoopAnimationsModule } from '@angular/platform-browser/animations';
import { By } from '@angular/platform-browser';
import { Store, StoreModule, combineReducers } from '@ngrx/store';
import * as EditorActions from '../../editor/editor.actions';
import * as fromRoot from '../../reducers';
import * as fromEditor from '../../editor/editor.reducer';
import { hot, cold } from 'jasmine-marbles';
import { MaterialModule } from 'src/app/shared';
import { DialogService } from 'src/app/core/services/dialog.service';
import { Article } from 'src/app/core';
import { Observable, of } from 'rxjs';

@Component({
    selector: 'app-list-error',
    template: ``
})
class TestHostListErrorComponent {
    @Input('errors') errors: any;
}

@Component({
    selector: 'app-editor-form',
    template: `
                <div id="submit" (click)="onSubmit()"></div>
                <div id="change" (click)="onWasChanged()"></div>`
})
class TestHostFrotmComponent {
    @Input('isSubmitting') isSubmitting: any;
    @Input('article') article: any;
    @Output() onsubmit = new EventEmitter<any>();
    @Output() wasChanged = new EventEmitter<boolean>();

    onSubmit() {
        this.onsubmit.emit(getArticle());
    }

    onWasChanged() {
        this.wasChanged.emit(true);
    }
}

describe('EditorComponent', () => {
    let component: EditorComponent;
    let fixture: ComponentFixture<EditorComponent>;
    let de: DebugElement;
    let store: Store<fromEditor.EditorState>;
    let article: Article;
    let dialogService: DialogService;

    class MockDialogService {
        confirmation = jest.fn();
    }


    beforeEach(
        async(() => {
            article = getArticle();
            const configure: ConfigureFn = testBed => {
                testBed.configureTestingModule({
                    declarations: [EditorComponent, TestHostFrotmComponent, TestHostListErrorComponent],
                    imports: [
                        MaterialModule,
                        NoopAnimationsModule,
                        StoreModule.forRoot({
                            ...fromRoot.reducers,
                            feature: combineReducers(fromEditor.editorReducer),
                        }),
                    ],
                    providers: [
                        {
                            provide: DialogService, useClass: MockDialogService
                        }
                    ]
                });
            };

            configureTests(configure).then(testBed => {
                fixture = testBed.createComponent(EditorComponent);
                component = fixture.componentInstance;
                de = fixture.debugElement;
                store = testBed.get(Store);
                fixture.detectChanges();

                dialogService = testBed.get(DialogService);

                spyOn(store, 'dispatch').and.callThrough();
            });
        })
    );

    it('should create', () => {
        expect(component).toBeTruthy();
    });

    it('should compile', () => {
        (<any>expect(fixture)).toMatchSnapshot();
    });

    it('should get article', () => {
        component.ngOnInit();

        const expected = cold('a', { a: null });

        expect(component.article$).toBeObservable(expected);
    });

    it('should get isSubmitting', () => {
        component.ngOnInit();

        const expected = cold('a', { a: false });

        expect(component.isSubmitting$).toBeObservable(expected);
    });

    it('should get errors', () => {
        component.ngOnInit();

        const expected = cold('a', { a: { errors: {} } });

        expect(component.errors$).toBeObservable(expected);
    });

    it('should dispatch "ClearEditorErrors"', () => {
        component.ngOnInit();
        expect(store.dispatch).toHaveBeenCalledWith(new EditorActions.ClearEditorErrors());
    });

    it('should handle ngOnDestroy', () => {
        component.ngOnDestroy();
        expect(store.dispatch).toHaveBeenCalledWith(new EditorActions.EditorArticleClear());
    });

    it('should dispatch "EditorArticleSaveRequest"', () => {
        const submitDe = de.query(By.css('#submit'));

        submitDe.triggerEventHandler('click', null);

        expect(store.dispatch).toHaveBeenCalledWith(new EditorActions.EditorArticleSaveRequest({ article }));
    });

    it('should set wasChanged to true', () => {
        const updateDe = de.query(By.css('#change'));

        updateDe.triggerEventHandler('click', null);

        expect(component.wasChanged).toBeTruthy();
    });

    it('shouldn\'t show dialog', () => {
        component.wasChanged = false;

        expect(component.canDeactivate()).toBeTruthy();
        expect(dialogService.confirmation).not.toHaveBeenCalled();
    });

    it('should show dialog and return true', () => {
        component.wasChanged = true;
        dialogService.confirmation = jest.fn(() => ({
            afterClosed() {
                return of(true);
            }
        }));

        (<Observable<boolean>>component.canDeactivate()).subscribe(res => {
            expect(res).toBeTruthy();
        });
        expect(dialogService.confirmation).toHaveBeenCalledWith({
            data: { question: 'You\'ll lost the data you have changed!' },
        });

    });

    it('should show dialog and return true', () => {
        component.wasChanged = true;
        dialogService.confirmation = jest.fn(() => ({
            afterClosed() {
                return of(false);
            }
        }));

        (<Observable<boolean>>component.canDeactivate()).subscribe(res => {
            expect(res).toBeFalsy();
        });
        expect(dialogService.confirmation).toHaveBeenCalledWith({
            data: { question: 'You\'ll lost the data you have changed!' },
        });

    });
});
