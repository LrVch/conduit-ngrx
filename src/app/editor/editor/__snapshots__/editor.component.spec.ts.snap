// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorComponent should compile 1`] = `
<app-editor
  article$={[Function Store]}
  dialog={[Function MockDialogService]}
  errors$={[Function Store]}
  isSubmitting$={[Function Store]}
  store={[Function Store]}
  translateService={[Function Object]}
  wasChanged="false"
>
  <div
    class="editor-page page"
  >
    <div
      class="container "
    >
      <div
        class="row"
      >
        <div
          class="col-md-10 offset-md-1 col-xs-12"
        >
          <app-list-error
            ng-reflect-errors="[object Object]"
          />
          <div
            class="mb-4"
          >
            <app-editor-form
              ng-reflect-is-submitting="false"
            >
              <div
                id="submit"
              />
              <div
                id="change"
              />
            </app-editor-form>
          </div>
        </div>
      </div>
    </div>
  </div>
</app-editor>
`;
